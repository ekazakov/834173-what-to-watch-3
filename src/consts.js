import PropTypes from "prop-types";

export const VIDEO_TIMER_HOVER = 1000;

export const SAME_FILMS_COUNT = 4;

export const SHOWN_FILMS_DEFAULT = 8;

export const Genres = {
  ALL: `All-genres`,
  HORROR: `horrors`,
};

export const TextCommentLength = {
  MIN: 40,
  MAX: 400,
};

export const MINUTES_IN_HOUR = 60;

export const filmProps = PropTypes.shape({
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  posterBig: PropTypes.string.isRequired,
  poster: PropTypes.string.isRequired,
  background: PropTypes.string.isRequired,
  backgroundColor: PropTypes.string.isRequired,
  video: PropTypes.string.isRequired,
  preview: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  genre: PropTypes.string.isRequired,
  rating: PropTypes.number.isRequired,
  score: PropTypes.number.isRequired,
  director: PropTypes.string.isRequired,
  starring: PropTypes.arrayOf(PropTypes.string).isRequired,
  duration: PropTypes.number.isRequired,
  year: PropTypes.number.isRequired,
  favorite: PropTypes.bool.isRequired,
});

export const commentProps = PropTypes.shape({
  id: PropTypes.number.isRequired,
  user: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
  }).isRequired,
  rating: PropTypes.number.isRequired,
  comment: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
});

export const filmsProps = PropTypes.arrayOf(filmProps).isRequired;
export const commentsProps = PropTypes.arrayOf(commentProps).isRequired;

export const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

export const ServerStatus = {
  AVAILABLE: `AVAILABLE`,
  NO_AVAILABLE: `NO_AVAILABLE`,
};

export const TabsName = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`,
};

export const AppRoute = {
  ROOT: `/`,
  LOGIN: `/login`,
  FAVORITE: `/mylist`,
  FILM: `/films`,
  REVIEW: `/review`,
  PLAYER: `/player`,
};

export const FavoriteStatus = {
  ADD: 1,
  DELETE: 0,
};
