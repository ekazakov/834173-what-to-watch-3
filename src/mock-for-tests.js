export const films = [
  {
    id: 1,
    name: `Harry Potter and the Goblet of Fire`,
    posterBig: `https://unsplash.it/280/175/`,
    poster: `https://unsplash.it/280/175/`,
    background: `https://unsplash.it/280/175/`,
    backgroundColor: `#ffffff`,
    video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    description: `Самый увлекательный в мире фильм`,
    genre: `drama`,
    rating: 6.7,
    score: 290,
    director: `Tarantino`,
    starring: [`Donatello`, `Rafael`, `Leonardo`],
    duration: 90,
    year: 2067,
    favorite: false,
  },
  {
    id: 2,
    name: `EuroTrip`,
    posterBig: `https://unsplash.it/280/175/`,
    poster: `https://unsplash.it/280/175/`,
    background: `https://unsplash.it/280/175/`,
    backgroundColor: `#ffffff`,
    video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    description: `Самый увлекательный в мире фильм`,
    genre: `drama`,
    rating: 6.7,
    score: 290,
    director: `Tarantino`,
    starring: [`Donatello`, `Rafael`, `Leonardo`],
    duration: 90,
    year: 2067,
    favorite: false,
  },
  {
    id: 3,
    name: `The Autopsy of Jane Doe`,
    posterBig: `https://unsplash.it/280/175/`,
    poster: `https://unsplash.it/280/175/`,
    background: `https://unsplash.it/280/175/`,
    backgroundColor: `#ffffff`,
    video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    description: `Самый увлекательный в мире фильм`,
    genre: `drama`,
    rating: 6.7,
    score: 290,
    director: `Tarantino`,
    starring: [`Donatello`, `Rafael`, `Leonardo`],
    duration: 90,
    year: 2067,
    favorite: false
  },
  {
    id: 4,
    name: `The Notebook`,
    posterBig: `https://unsplash.it/280/175/`,
    poster: `https://unsplash.it/280/175/`,
    background: `https://unsplash.it/280/175/`,
    backgroundColor: `#ffffff`,
    video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    description: `Самый увлекательный в мире фильм`,
    genre: `drama`,
    rating: 6.7,
    score: 290,
    director: `Tarantino`,
    starring: [`Donatello`, `Rafael`, `Leonardo`],
    duration: 90,
    year: 2067,
    favorite: false
  },
  {
    id: 5,
    name: `Carri`,
    posterBig: `https://unsplash.it/280/175/`,
    poster: `https://unsplash.it/280/175/`,
    background: `https://unsplash.it/280/175/`,
    backgroundColor: `#ffffff`,
    video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    description: `Самый увлекательный в мире фильм`,
    genre: `drama`,
    rating: 6.7,
    score: 290,
    director: `Tarantino`,
    starring: [`Donatello`, `Rafael`, `Leonardo`],
    duration: 90,
    year: 2067,
    favorite: false
  },
];

export const comments = [
  {
    id: 0,
    user: {
      id: 0,
      name: `Kate`,
    },
    rating: 7.6,
    comment: `OMG`,
    date: `122`,
  },
  {
    id: 1,
    user: {
      id: 0,
      name: `David`,
    },
    rating: 7.6,
    comment: `OMG`,
    date: `122`,
  },
];
