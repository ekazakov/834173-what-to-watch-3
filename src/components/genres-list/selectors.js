import {Genres} from "../../consts.js";

export const getGenre = (state) => {
  return state.genre;
};

export const getGenres = (state) => {
  return [Genres.ALL, ...new Set(state.films.map((film) => film.genre))];
};

export const getFilteredFilms = (state) => {
  return getGenre(state) === Genres.ALL ? state.films : state.films.filter((film) => film.genre === getGenre(state));
};
