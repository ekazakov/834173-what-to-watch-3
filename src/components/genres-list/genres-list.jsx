import React from "react";
import {connect} from "react-redux";
import PropTypes from "prop-types";
import {ActionCreator} from "../../reducer/state/state.js";
import FilmsList from "../films-list/films-list.jsx";
import {filmsProps, SHOWN_FILMS_DEFAULT} from "../../consts.js";
import {getGenre, getGenres, getLimitedFilms} from "../../reducer/state/selectors.js";

const GenresList = (props) => {
  const {genre, genres, changeGenre, onFilmClick, resetFilmsAmount, limitedFilms} = props;

  const onGenreClick = (availableGenre) => {
    changeGenre(availableGenre);
    resetFilmsAmount(SHOWN_FILMS_DEFAULT);
  };

  return (
    <React.Fragment>
      <ul className="catalog__genres-list">
        {genres.map((availableGenre, index) => (
          <li
            className={`catalog__genres-item ${genre === availableGenre ? `catalog__genres-item--active` : ``}`}
            key={availableGenre + index}>
            <a className="catalog__genres-link" onClick={() => onGenreClick(availableGenre)}>
              {availableGenre}
            </a>
          </li>
        ))}
      </ul>

      <FilmsList
        films={limitedFilms}
        onFilmClick={onFilmClick}
      />
    </React.Fragment>
  );
};

GenresList.propTypes = {
  genre: PropTypes.string.isRequired,
  changeGenre: PropTypes.func.isRequired,
  genres: PropTypes.array.isRequired,
  onFilmClick: PropTypes.func.isRequired,
  resetFilmsAmount: PropTypes.func.isRequired,
  limitedFilms: filmsProps,
};

const mapStateToProps = (state) => ({
  genre: getGenre(state),
  genres: getGenres(state),
  limitedFilms: getLimitedFilms(state),
});

const mapDispatchToProps = (dispatch) => ({
  changeGenre(genre) {
    dispatch(ActionCreator.changeGenre(genre));
  },
  resetFilmsAmount(shownFilms) {
    dispatch(ActionCreator.resetFilmsAmount(shownFilms));
  },
});

export {GenresList};
export default connect(mapStateToProps, mapDispatchToProps)(GenresList);
