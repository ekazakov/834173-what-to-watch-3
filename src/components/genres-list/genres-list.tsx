import * as React from "react";
import {connect} from "react-redux";
import FilmsList from "../films-list/films-list";
import {ActionCreator} from "../../reducer/state/state";
import {getGenre, getGenres, getLimitedFilms} from "../../reducer/state/selectors";
import {SHOWN_FILMS_DEFAULT} from "../../consts";
import {Film} from "../../types";

interface GenresListProps {
  genre: string,
  changeGenre: (string) => void,
  genres: string[],
  onFilmClick: () => void,
  resetFilmsAmount: (number) => void,
  limitedFilms: Film[],
}

const GenresList: React.FunctionComponent<GenresListProps> = (props: GenresListProps) => {
  const {genre, genres, changeGenre, onFilmClick, resetFilmsAmount, limitedFilms} = props;

  const onGenreClick = (availableGenre) => {
    changeGenre(availableGenre);
    resetFilmsAmount(SHOWN_FILMS_DEFAULT);
  };

  return (
    <React.Fragment>
      <ul className="catalog__genres-list">
        {genres.map((availableGenre, index) => (
          <li
            className={`catalog__genres-item ${genre === availableGenre ? `catalog__genres-item--active` : ``}`}
            key={availableGenre + index}>
            <a className="catalog__genres-link" onClick={() => onGenreClick(availableGenre)}>
              {availableGenre}
            </a>
          </li>
        ))}
      </ul>

      <FilmsList
        films={limitedFilms}
        onFilmClick={onFilmClick}
      />
    </React.Fragment>
  );
};

const mapStateToProps = (state) => ({
  genre: getGenre(state),
  genres: getGenres(state),
  limitedFilms: getLimitedFilms(state),
});

const mapDispatchToProps = (dispatch) => ({
  changeGenre(genre) {
    dispatch(ActionCreator.changeGenre(genre));
  },
  resetFilmsAmount(shownFilms) {
    dispatch(ActionCreator.resetFilmsAmount(shownFilms));
  },
});

export {GenresList};
export default connect(mapStateToProps, mapDispatchToProps)(GenresList);
