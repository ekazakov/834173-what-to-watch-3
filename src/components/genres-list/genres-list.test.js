import React from "react";
import renderer from "react-test-renderer";
import {GenresList} from "./genres-list.jsx";
import {Genres} from "../../consts.js";
import {MemoryRouter} from "react-router-dom";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import NameSpace from "../../reducer/name-space";
import {AuthorizationStatus, SHOWN_FILMS_DEFAULT, TabsName} from "../../consts";
import {films, comments, genres} from "../../mock-for-tests";

const mockStore = configureStore([]);

it(`Should GenresList render correctly`, () => {
  const store = mockStore({
    [NameSpace.STATE]: {
      genre: Genres.ALL,
      chosenFilmId: films[0].id,
      currentTab: TabsName.OVERVIEW,
    },
    [NameSpace.DATA]: {
      films,
      comments,
    },
    [NameSpace.USER]: {
      authorizationStatus: AuthorizationStatus.NO_AUTH,
    },
  });


  const tree = renderer.create(
      <MemoryRouter>
        <Provider store={store}>
          <GenresList
            shownFilms={SHOWN_FILMS_DEFAULT}
            genre={Genres.ALL}
            genres={genres}
            changeGenre={() => {}}
            onTitleOfFilmClick={() => {}}
            resetFilmsAmount={() => {}}
            filteredFilms={films}
          />
        </Provider>
      </MemoryRouter>, {
        createNodeMock: () => {
          return {};
        }
      }
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
