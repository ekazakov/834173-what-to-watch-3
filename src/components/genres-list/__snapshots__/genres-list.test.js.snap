// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should GenresList render correctly 1`] = `
Array [
  <ul
    className="catalog__genres-list"
  >
    <li
      className="catalog__genres-item catalog__genres-item--active"
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        All-genres
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        drama
      </a>
    </li>
  </ul>,
  <div
    className="catalog__movies-list"
  >
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <video
          height="175"
          muted={true}
          width="100%"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/dev-film"
          onClick={[Function]}
        >
          Harry Potter and the Goblet of Fire
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <video
          height="175"
          muted={true}
          width="100%"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/dev-film"
          onClick={[Function]}
        >
          EuroTrip
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <video
          height="175"
          muted={true}
          width="100%"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/dev-film"
          onClick={[Function]}
        >
          The Autopsy of Jane Doe
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <video
          height="175"
          muted={true}
          width="100%"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/dev-film"
          onClick={[Function]}
        >
          The Notebook
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <video
          height="175"
          muted={true}
          width="100%"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/dev-film"
          onClick={[Function]}
        >
          Carri
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <video
          height="175"
          muted={true}
          width="100%"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/dev-film"
          onClick={[Function]}
        >
          Carri
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <video
          height="175"
          muted={true}
          width="100%"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/dev-film"
          onClick={[Function]}
        >
          Carri
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <video
          height="175"
          muted={true}
          width="100%"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/dev-film"
          onClick={[Function]}
        >
          Carri
        </a>
      </h3>
    </article>
  </div>,
]
`;
